mvpa2/algorithms/group_clusterthr.py

```diff
@@ -0,0 +1,427 @@
+            doc="""Feature-wise probability threshold. The value corresponding
+            to this probability in the NULL distribution of accuracies will
+            be used as threshold for cluster forming. Given that the NULL
+            distribution is estimated per feature, the actual threshold value
+            will vary across features yielding a threshold vector. The number
+            of bootstrap samples need to be adequate for a desired probability.
+            A ``ValueError`` is raised otherwise.""")
+
+    chunk_attr = Parameter('chunks',
+            doc="""Name of the attribute indicating the individual chunks from
+            which a single sample each is drawn for averaging into a bootstrap
+            sample.""")
+
+    fwe_rate = Parameter(0.05,
+            constraints=EnsureFloat() & EnsureRange(min=0.0, max=1.0),
+            doc="""Family-wise error rate for multiple comparison correction
+            of cluster size probabilities.""")
+
+    multicomp_correction = Parameter('fdr_bh',
+             constraints=EnsureChoice('bonferroni', 'sidak', 'holm-sidak',
+                                      'holm', 'simes-hochberg', 'hommel',
+                                      'fdr_bh', 'fdr_by', None),
+             doc="""Strategy for multiple comparison correction of cluster
+             probailities. All methods supported by statsmodels' ``multitest``
+             are available. In addition, ``None`` can be specific to disable
+             correction.""")
+
+    n_blocks = Parameter(1, constraints=EnsureInt() & EnsureRange(min=1),
+             doc="""Number of segments used to compute the feature-wise NULL
+             distributions. This parameter determines the peak memory demand.
+             In case of a single segment a matrix of size
+             (n_bootstrap x nfeatures) will be allocated. Increasing the number
+             of segments reduces the peak memory demand by that roughly factor.
+             """)
+
+    def __init__(self, **kwargs):
+        # force disable auto-train: would make no sense
+        Learner.__init__(self, auto_train=False, **kwargs)
+        self.untrain()
+
+    def _untrain(self):
+        self._thrmap = None
+        self._null_cluster_sizes = None
+
+    def _train(self, ds):
+        # shortcuts
+        chunk_attr = self.params.chunk_attr
+        #
+        # Step 0: bootstrap maps by drawing one for each chunk and average them
+        # (do N iterations)
+        # this could take a lot of memory, hence instead of computing the maps
+        # we compute the source maps they can be computed from and then (re)build
+        # the matrix of bootstrapped maps either row-wise or column-wise (as
+        # needed) to save memory by a factor of (close to) `n_bootstrap`
+        # which samples belong to which chunk
+        chunk_samples = dict([(c, np.where(ds.sa[chunk_attr].value == c)[0])
+                                    for c in ds.sa[chunk_attr].unique])
```

Comment:
You can find examples in this code file where this kind of suggested indentation would be incomprehensible (nested list comprehensions). Unless you force me, I'd like to keep it this way.